File: src-tauri/src/site.rs
use anyhow::{anyhow, bail, Context, Result};
use percent_encoding::percent_decode_str;
use reqwest::header::{HeaderMap, HeaderValue};
use reqwest::Client;
use scraper::{ElementRef, Html, Selector};
use serde::{Deserialize, Serialize};

pub const ORIGIN: &str = "https://animejoya.ru";
pub const UA: &str = "Mozilla/5.0 (X11; Linux x86_64; rv:133.0) Gecko/20100101 Firefox/133.0";

#[derive(Debug, Clone, Serialize)]
pub struct Source {
    pub quality: String,
    pub url: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct Episode {
    pub player_id: String,
    pub title: String,
    pub sources: Vec<Source>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Player {
    pub id: String,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Fact {
    pub key: String,
    pub value: String,
}

/// Шапка страницы тайтла: постер, описание, таблица характеристик.
#[derive(Debug, Clone, Serialize, Default)]
pub struct Meta {
    pub title: String,
    pub original: String,
    pub poster: String,
    pub description: String,
    pub genres: Vec<String>,
    pub facts: Vec<Fact>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Playlist {
    pub players: Vec<Player>,
    pub episodes: Vec<Episode>,
}

impl Playlist {
    /// data-id иерархичен: `0_0` — плеер, `0_1_0` — озвучка `0_1` + плеер `_0`.
    pub fn player_name(&self, id: &str) -> String {
        let parts: Vec<&str> = id.split('_').collect();
        let chain: Vec<String> = (1..=parts.len())
            .map(|n| parts[..n].join("_"))
            .filter_map(|prefix| self.label(&prefix))
            .collect();
        if chain.is_empty() {
            id.to_string()
        } else {
            chain.join(" · ")
        }
    }

    fn label(&self, id: &str) -> Option<String> {
        self.players
            .iter()
            .find(|p| p.id == id)
            .map(|p| p.name.clone())
    }

    /// Плееры, отдающие прямые ссылки на видео (а не чужой iframe).
    pub fn direct_players(&self) -> Vec<Player> {
        let mut ids: Vec<String> = Vec::new();
        for ep in &self.episodes {
            if !ep.sources.is_empty() && !ids.contains(&ep.player_id) {
                ids.push(ep.player_id.clone());
            }
        }
        ids.into_iter()
            .map(|id| Player {
                name: self.player_name(&id),
                id,
            })
            .collect()
    }

    pub fn external_names(&self) -> Vec<String> {
        let mut out: Vec<String> = Vec::new();
        for ep in &self.episodes {
            if ep.sources.is_empty() {
                let leaf = self
                    .label(&ep.player_id)
                    .unwrap_or_else(|| ep.player_id.clone());
                if !out.contains(&leaf) {
                    out.push(leaf);
                }
            }
        }
        out
    }

    pub fn episodes_of(&self, player_id: &str) -> Vec<&Episode> {
        self.episodes
            .iter()
            .filter(|e| e.player_id == player_id)
            .collect()
    }
}

pub struct Site {
    pub http: Client,
}

impl Site {
    pub fn new() -> Result<Self> {
        let mut h = HeaderMap::new();
        h.insert(
            "Accept-Language",
            HeaderValue::from_static("ru-RU,ru;q=0.9,en-US;q=0.7,en;q=0.6"),
        );
        h.insert("DNT", HeaderValue::from_static("1"));
        h.insert("Upgrade-Insecure-Requests", HeaderValue::from_static("1"));

        let http = Client::builder()
            .user_agent(UA)
            .default_headers(h)
            .cookie_store(true)
            .gzip(true)
            .brotli(true)
            .timeout(std::time::Duration::from_secs(60))
            .build()?;
        Ok(Self { http })
    }

    /// GET страницы «как из браузера».
    pub async fn get_page(&self, url: &str) -> Result<String> {
        let r = self
            .http
            .get(url)
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
            .header("Sec-Fetch-Dest", "document")
            .header("Sec-Fetch-Mode", "navigate")
            .header("Sec-Fetch-Site", "none")
            .header("Sec-Fetch-User", "?1")
            .send()
            .await
            .with_context(|| format!("не удалось открыть {url}"))?;
        // Страница без прав доступа отдаётся с кодом 403, но с нужным HTML.
        Ok(r.text().await?)
    }

    pub async fn login(&self, user: &str, pass: &str, referer: &str) -> Result<()> {
        let form = [
            ("login_name", user),
            ("login_password", pass),
            ("login_not_save", "0"),
            ("login", "submit"),
        ];
        let r = self
            .http
            .post(referer)
            .header(
                "Accept",
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            )
            .header("Origin", ORIGIN)
            .header("Referer", referer)
            .header("Sec-Fetch-Dest", "document")
            .header("Sec-Fetch-Mode", "navigate")
            .header("Sec-Fetch-Site", "same-origin")
            .header("Sec-Fetch-User", "?1")
            .form(&form)
            .send()
            .await
            .context("запрос авторизации не прошёл")?;
        let body = r.text().await?;
        if body.contains("name=\"login_name\"") {
            bail!("не удалось войти — проверьте логин и пароль");
        }
        Ok(())
    }

    /// Плейлист приходит отдельным ajax-запросом, в HTML страницы его нет.
    pub async fn playlist(&self, news_id: &str, referer: &str) -> Result<Playlist> {
        let url = format!("{ORIGIN}/engine/ajax/playlists.php?news_id={news_id}&xfield=playlist");
        let r = self
            .http
            .get(&url)
            .header("Accept", "application/json, text/javascript, */*; q=0.01")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Referer", referer)
            .header("Sec-Fetch-Dest", "empty")
            .header("Sec-Fetch-Mode", "cors")
            .header("Sec-Fetch-Site", "same-origin")
            .send()
            .await
            .context("не удалось получить плейлист")?;

        let json: serde_json::Value = r.json().await.context("плейлист вернулся не в JSON")?;
        if json.get("success").and_then(|v| v.as_bool()) != Some(true) {
            bail!("сайт не отдал плейлист (нет доступа к этой странице?)");
        }
        let html = json
            .get("response")
            .and_then(|v| v.as_str())
            .ok_or_else(|| anyhow!("в ответе нет поля response"))?;

        Ok(parse_playlist(html))
    }
}

fn parse_playlist(html: &str) -> Playlist {
    let doc = Html::parse_fragment(html);
    let sel_players = Selector::parse(".playlists-lists li[data-id]").unwrap();
    let sel_eps = Selector::parse(".playlists-videos li[data-file]").unwrap();

    let players = doc
        .select(&sel_players)
        .filter_map(|el| {
            Some(Player {
                id: el.value().attr("data-id")?.to_string(),
                name: text_of(&el),
            })
        })
        .collect();

    let episodes = doc
        .select(&sel_eps)
        .filter_map(|el| {
            let file = el.value().attr("data-file")?;
            Some(Episode {
                player_id: el.value().attr("data-id").unwrap_or("?").to_string(),
                title: text_of(&el),
                sources: parse_sources(file),
            })
        })
        .collect();

    Playlist { players, episodes }
}

fn text_of(el: &ElementRef) -> String {
    el.text()
        .collect::<String>()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
}

/// `//animejoya.ru/player/playerjs.html?skip=..&file=[1080p]https://..a.mp4,[720p]https://..b.mp4`
fn parse_sources(data_file: &str) -> Vec<Source> {
    if !data_file.contains("playerjs.html") {
        return Vec::new();
    }
    let Some(idx) = data_file.find("file=") else {
        return Vec::new();
    };
    let raw = &data_file[idx + "file=".len()..];
    let list = percent_decode_str(raw).decode_utf8_lossy().to_string();

    // Границы записей — `[` в начале строки или сразу после запятой.
    let marks: Vec<usize> = list
        .match_indices('[')
        .map(|(i, _)| i)
        .filter(|&i| i == 0 || list[..i].ends_with(','))
        .collect();

    if marks.is_empty() {
        let url = list.trim().to_string();
        return if url.starts_with("http") {
            vec![Source {
                quality: "video".into(),
                url,
            }]
        } else {
            Vec::new()
        };
    }

    let mut out = Vec::new();
    for (n, &start) in marks.iter().enumerate() {
        let end = marks.get(n + 1).map(|&e| e - 1).unwrap_or(list.len());
        let entry = &list[start..end];
        let Some(close) = entry.find(']') else {
            continue;
        };
        let url = entry[close + 1..].trim().to_string();
        if url.starts_with("http") {
            out.push(Source {
                quality: entry[1..close].to_string(),
                url,
            });
        }
    }
    out
}

/// news_id нужен для ajax-запроса плейлиста.
pub fn parse_news_id(html: &str) -> Option<String> {
    let i = html.find("data-news_id=")?;
    let rest = &html[i + "data-news_id=".len()..];
    let rest = rest.trim_start_matches(['"', '\'']);
    let id: String = rest.chars().take_while(|c| c.is_ascii_digit()).collect();
    (!id.is_empty()).then_some(id)
}

pub fn parse_meta(html: &str) -> Meta {
    let doc = Html::parse_document(html);
    let one = |css: &str| -> Option<String> {
        Selector::parse(css)
            .ok()
            .and_then(|s| doc.select(&s).next().map(|el| text_of(&el)))
    };
    let attr = |css: &str, name: &str| -> Option<String> {
        Selector::parse(css)
            .ok()
            .and_then(|s| doc.select(&s).next())
            .and_then(|el| el.value().attr(name).map(str::to_string))
    };

    let mut facts = Vec::new();
    let mut genres = Vec::new();
    if let Ok(sel) = Selector::parse(".blkdesc p") {
        let label = Selector::parse(".timpact").unwrap();
        let link = Selector::parse("a").unwrap();
        for p in doc.select(&sel) {
            let Some(lab) = p.select(&label).next() else {
                continue;
            };
            let key = text_of(&lab).trim_end_matches(':').to_string();
            let full = text_of(&p);
            let value = full
                .strip_prefix(&text_of(&lab))
                .unwrap_or(&full)
                .trim()
                .to_string();
            if key.eq_ignore_ascii_case("жанр") {
                genres = p.select(&link).map(|a| text_of(&a)).collect();
            }
            if !value.is_empty() {
                facts.push(Fact { key, value });
            }
        }
    }

    Meta {
        title: one("h1").unwrap_or_else(|| "Без названия".into()),
        original: one("h2.romanji").unwrap_or_default(),
        poster: attr("meta[property=\"og:image\"]", "content").unwrap_or_default(),
        description: one("[itemprop=description]")
            .map(|d| d.trim_start_matches("Описание:").trim().to_string())
            .or_else(|| attr("meta[property=\"og:description\"]", "content"))
            .unwrap_or_default(),
        genres,
        facts,
    }
}

/// У гостя страница отдаётся с формой входа и без плейлиста.
pub fn is_authorized(html: &str) -> bool {
    !html.contains("name=\"login_name\"") && parse_news_id(html).is_some()
}

/// Ссылки на видео лежат на CDN и берутся уже без кук, но с браузерными заголовками.
pub fn media_client() -> Result<Client> {
    Ok(Client::builder()
        .user_agent(UA)
        .connect_timeout(std::time::Duration::from_secs(30))
        .build()?)
}

File: server/site.ts
import * as cheerio from "cheerio";
import type { AnyNode } from "domhandler";
import type { Fact } from "./store.ts";

export const SIBNET = "https://video.sibnet.ru";
export const UA = "Mozilla/5.0 (X11; Linux x86_64; rv:133.0) Gecko/20100101 Firefox/133.0";

export type Source = { quality: string; url: string; referer: string };
export type Episode = { playerId: string; title: string; sources: Source[]; embed: string };
export type Player = { id: string; name: string };
export type Choice = { id: string; path: string[]; resolvable: boolean };
export type Meta = {
  title: string;
  original: string;
  poster: string;
  description: string;
  genres: string[];
  facts: Fact[];
};

export type Kind = "direct" | "allvideo" | "sibnet" | "external";

export function kind(dataFile: string): Kind {
  if (dataFile.includes("playerjs.html")) return "direct";
  if (dataFile.includes("fsst.online")) return "allvideo";
  if (dataFile.includes("sibnet.ru")) return "sibnet";
  return "external";
}

/// `true` — ссылки достаются отдельным запросом, `null` — не умеем вовсе.
function usable(ep: Episode): boolean | null {
  switch (kind(ep.embed)) {
    case "direct":
      return ep.sources.length > 0 ? false : null;
    case "allvideo":
    case "sibnet":
      return true;
    default:
      return null;
  }
}

const absolute = (url: string): string => (url.startsWith("//") ? `https:${url}` : url);

/// У сайта несколько доменов-зеркал — всё строим от того, с которого пришли.
const originOf = (url: string): string => new URL(url).origin;

const BASE_HEADERS: Record<string, string> = {
  "User-Agent": UA,
  "Accept-Language": "ru-RU,ru;q=0.9,en-US;q=0.7,en;q=0.6",
  DNT: "1",
  "Upgrade-Insecure-Requests": "1",
};

/// Минимальная банка кук: сайт один, поэтому хватает пары имя-значение на хост.
class Jar {
  private byHost = new Map<string, Map<string, string>>();

  read(url: string): string {
    const jar = this.byHost.get(new URL(url).hostname);
    if (!jar || jar.size === 0) return "";
    return [...jar].map(([k, v]) => `${k}=${v}`).join("; ");
  }

  write(url: string, res: Response): void {
    const set = res.headers.getSetCookie();
    if (set.length === 0) return;
    const host = new URL(url).hostname;
    const jar = this.byHost.get(host) ?? new Map<string, string>();
    for (const line of set) {
      const pair = line.split(";")[0] ?? "";
      const i = pair.indexOf("=");
      if (i > 0) jar.set(pair.slice(0, i).trim(), pair.slice(i + 1).trim());
    }
    this.byHost.set(host, jar);
  }
}

type Init = { method?: string; headers: Record<string, string>; body?: string };

/// Редиректы ведём вручную: иначе куки, выставленные на промежуточном 302
/// (а именно так приходит сессия после логина), до нас не доедут.
async function hop(start: string, init: Init, jar: Jar, timeoutMs: number): Promise<Response> {
  let url = start;
  let opts = { ...init };
  for (let n = 0; n < 10; n++) {
    const cookie = jar.read(url);
    const headers = { ...BASE_HEADERS, ...opts.headers, ...(cookie ? { Cookie: cookie } : {}) };
    const res = await fetch(url, {
      method: opts.method ?? "GET",
      headers,
      body: opts.body,
      redirect: "manual",
      signal: AbortSignal.timeout(timeoutMs),
    });
    jar.write(url, res);
    const loc = res.headers.get("location");
    if (res.status >= 300 && res.status < 400 && loc) {
      const next = new URL(loc, url).toString();
      await res.body?.cancel();
      if (res.status === 303 || (opts.method === "POST" && res.status !== 307 && res.status !== 308)) {
        opts = { headers: opts.headers };
      }
      url = next;
      continue;
    }
    return res;
  }
  throw new Error("слишком много редиректов");
}

export class Site {
  private jar = new Jar();

  /// GET страницы «как из браузера».
  async getPage(url: string): Promise<string> {
    let res: Response;
    try {
      res = await hop(
        url,
        {
          headers: {
            Accept:
              "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
            "Sec-Fetch-Dest": "document",
            "Sec-Fetch-Mode": "navigate",
            "Sec-Fetch-Site": "none",
            "Sec-Fetch-User": "?1",
          },
        },
        this.jar,
        60_000,
      );
    } catch (e) {
      throw new Error(`не удалось открыть ${url}: ${e instanceof Error ? e.message : e}`);
    }
    // Страница без прав доступа отдаётся с кодом 403, но с нужным HTML.
    return res.text();
  }

  async login(user: string, pass: string, referer: string): Promise<void> {
    const form = new URLSearchParams({
      login_name: user,
      login_password: pass,
      login_not_save: "0",
      login: "submit",
    });
    let res: Response;
    try {
      res = await hop(
        referer,
        {
          method: "POST",
          headers: {
            Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Content-Type": "application/x-www-form-urlencoded",
            Origin: originOf(referer),
            Referer: referer,
            "Sec-Fetch-Dest": "document",
            "Sec-Fetch-Mode": "navigate",
            "Sec-Fetch-Site": "same-origin",
            "Sec-Fetch-User": "?1",
          },
          body: form.toString(),
        },
        this.jar,
        60_000,
      );
    } catch (e) {
      throw new Error(`запрос авторизации не прошёл: ${e instanceof Error ? e.message : e}`);
    }
    const body = await res.text();
    if (body.includes('name="login_name"')) {
      throw new Error("не удалось войти — проверьте логин и пароль");
    }
  }

  /// Страница чужого плеера — берём её так, как её брал бы iframe на сайте.
  private async embedPage(url: string, from: string): Promise<string> {
    let res: Response;
    try {
      res = await hop(
        url,
        {
          headers: {
            Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            Referer: `${from}/`,
            "Sec-Fetch-Dest": "iframe",
            "Sec-Fetch-Mode": "navigate",
            "Sec-Fetch-Site": "cross-site",
          },
        },
        this.jar,
        60_000,
      );
    } catch (e) {
      throw new Error(`не удалось открыть плеер ${url}: ${e instanceof Error ? e.message : e}`);
    }
    return res.text();
  }

  /// Прямые ссылки для одной серии внешнего плеера.
  async resolve(embed: string, from: string): Promise<Source[]> {
    const url = absolute(embed);
    switch (kind(embed)) {
      case "direct":
        return parseSources(embed);
      case "allvideo":
        return this.allvideo(url, from);
      case "sibnet":
        return this.sibnet(url, from);
      default:
        throw new Error(`плеер ${url} не поддерживается`);
    }
  }

  /// Формат `[качество]ссылка` у fsst.online тот же, что и у своего плеера сайта.
  /// Referer оставляем пустым: CDN за ссылкой пускает только свой домен.
  private async allvideo(url: string, from: string): Promise<Source[]> {
    const html = await this.embedPage(url, from);
    const list = playerjsFile(html);
    const out = list ? parseQualityList(list) : [];
    if (out.length === 0) throw new Error("AllVideo не отдал ссылок на видео");
    return out;
  }

  /// Sibnet прячет видео за редиректом, который без своего Referer даёт 403.
  /// Ссылку `/v/..` не разворачиваем: она вечная, а вот CDN за ней — на пару часов.
  private async sibnet(url: string, from: string): Promise<Source[]> {
    const html = await this.embedPage(url, from);
    const p = sibnetPath(html);
    if (!p) throw new Error("Sibnet не отдал ссылки на видео");
    return [
      {
        quality: "sibnet",
        url: p.startsWith("http") ? p : `${SIBNET}${p}`,
        referer: `${SIBNET}/`,
      },
    ];
  }

  /// Плейлист приходит отдельным ajax-запросом, в HTML страницы его нет.
  async playlist(newsId: string, referer: string): Promise<Playlist> {
    const url = `${originOf(referer)}/engine/ajax/playlists.php?news_id=${newsId}&xfield=playlist`;
    let res: Response;
    try {
      res = await hop(
        url,
        {
          headers: {
            Accept: "application/json, text/javascript, */*; q=0.01",
            "X-Requested-With": "XMLHttpRequest",
            Referer: referer,
            "Sec-Fetch-Dest": "empty",
            "Sec-Fetch-Mode": "cors",
            "Sec-Fetch-Site": "same-origin",
          },
        },
        this.jar,
        60_000,
      );
    } catch (e) {
      throw new Error(`не удалось получить плейлист: ${e instanceof Error ? e.message : e}`);
    }
    let json: { success?: boolean; response?: string };
    try {
      json = (await res.json()) as typeof json;
    } catch {
      throw new Error("плейлист вернулся не в JSON");
    }
    if (json.success !== true) throw new Error("сайт не отдал плейлист (нет доступа к этой странице?)");
    if (typeof json.response !== "string") throw new Error("в ответе нет поля response");
    return parsePlaylist(json.response);
  }
}

export class Playlist {
  constructor(
    readonly players: Player[],
    readonly episodes: Episode[],
  ) {}

  /// data-id иерархичен: `0_0` — озвучка, `0_0_1` — плеер внутри неё,
  /// а у долгих тайтлов есть ещё `0_0_1_3` — диапазон серий.
  playerPath(id: string): string[] {
    const parts = id.split("_");
    const chain: string[] = [];
    for (let n = 1; n <= parts.length; n++) {
      const label = this.label(parts.slice(0, n).join("_"));
      if (label !== null) chain.push(label);
    }
    return chain.length > 0 ? chain : [id];
  }

  private label(id: string): string | null {
    return this.players.find((p) => p.id === id)?.name ?? null;
  }

  /// Плееры, из которых достаём прямые ссылки. Флаг — нужен ли для этого
  /// отдельный запрос к чужому сайту (AllVideo, Sibnet).
  playable(): Choice[] {
    const out: Choice[] = [];
    for (const ep of this.episodes) {
      const resolvable = usable(ep);
      if (resolvable === null) continue;
      if (out.some((c) => c.id === ep.playerId)) continue;
      out.push({ id: ep.playerId, path: this.playerPath(ep.playerId), resolvable });
    }
    return out;
  }

  externalNames(): string[] {
    const out: string[] = [];
    for (const ep of this.episodes) {
      if (usable(ep) !== null) continue;
      // Название плеера — второй уровень: [озвучка, плеер, диапазон серий].
      const path = this.playerPath(ep.playerId);
      const name = path[1] ?? path[path.length - 1];
      if (name === undefined) continue;
      if (!out.includes(name)) out.push(name);
    }
    return out;
  }

  episodesOf(playerId: string): Episode[] {
    return this.episodes.filter((e) => e.playerId === playerId);
  }
}

type Api = cheerio.CheerioAPI;

const textOf = ($: Api, el: AnyNode): string => $(el).text().split(/\s+/).filter(Boolean).join(" ");

export function parsePlaylist(html: string): Playlist {
  const $ = cheerio.load(html, null, false);

  const players: Player[] = [];
  $(".playlists-lists li[data-id]").each((_, el) => {
    const id = $(el).attr("data-id");
    if (id !== undefined) players.push({ id, name: textOf($, el) });
  });

  const episodes: Episode[] = [];
  $(".playlists-videos li[data-file]").each((_, el) => {
    const file = $(el).attr("data-file");
    if (file === undefined) return;
    episodes.push({
      playerId: $(el).attr("data-id") ?? "?",
      title: textOf($, el),
      sources: parseSources(file),
      embed: file,
    });
  });

  return new Playlist(players, episodes);
}

/// `//animejoya.ru/player/playerjs.html?file=[1080p]https://..a.mp4,[720p]https://..b.mp4&skip=..`
export function parseSources(dataFile: string): Source[] {
  if (!dataFile.includes("playerjs.html")) return [];
  const q = dataFile.indexOf("?");
  if (q < 0) return [];
  // `skip` встречается и до, и после `file`, поэтому режем по границе параметра.
  for (const pair of dataFile.slice(q + 1).split("&")) {
    if (!pair.startsWith("file=")) continue;
    const raw = pair.slice("file=".length);
    let list: string;
    try {
      list = decodeURIComponent(raw);
    } catch {
      list = raw;
    }
    return parseQualityList(list);
  }
  return [];
}

/// `[1080p]https://a.mp4,[720p]https://b.mp4` — общий формат PlayerJS.
export function parseQualityList(list: string): Source[] {
  // Границы записей — `[` в начале строки или сразу после запятой.
  const marks: number[] = [];
  for (let i = 0; i < list.length; i++) {
    if (list[i] === "[" && (i === 0 || list[i - 1] === ",")) marks.push(i);
  }

  if (marks.length === 0) {
    const url = list.trim();
    return url.startsWith("http") ? [{ quality: "video", url, referer: "" }] : [];
  }

  const out: Source[] = [];
  for (let n = 0; n < marks.length; n++) {
    const start = marks[n]!;
    const end = n + 1 < marks.length ? marks[n + 1]! - 1 : list.length;
    const entry = list.slice(start, end);
    const close = entry.indexOf("]");
    if (close < 0) continue;
    const url = entry.slice(close + 1).trim();
    if (url.startsWith("http")) out.push({ quality: entry.slice(1, close), url, referer: "" });
  }
  return out;
}

/// Значение JS-строки следом за `key`: `file:"..."`, `src: '...'`.
/// Возвращает ещё и позицию за ней, чтобы можно было искать дальше.
function jsStringAfter(hay: string, key: string, from: number): [string, number] | null {
  const found = hay.indexOf(key, from);
  if (found < 0) return null;
  const at = found + key.length;
  const rest = hay.slice(at);
  const open = rest.search(/["']/);
  if (open < 0) return null;
  const quote = rest[open]!;
  const start = open + 1;
  const end = rest.indexOf(quote, start);
  if (end < 0) return null;
  return [rest.slice(start, end), at + end];
}

/// Первый `file:` в конфиге PlayerJS, похожий на список ссылок.
function playerjsFile(html: string): string | null {
  let at = 0;
  for (;;) {
    const hit = jsStringAfter(html, "file:", at);
    if (hit === null) return null;
    const [value, next] = hit;
    if (value.startsWith("[") || value.startsWith("http")) return value;
    at = next;
  }
}

/// `player.src([{src: "/v/<hash>/<id>.mp4", type: "video/mp4"},]);`
function sibnetPath(html: string): string | null {
  const at = html.indexOf("player.src(");
  if (at < 0) return null;
  const hit = jsStringAfter(html, "src", at);
  if (hit === null) return null;
  const [value] = hit;
  return value.includes(".mp4") || value.includes(".m3u8") ? value : null;
}

/// news_id нужен для ajax-запроса плейлиста.
export function parseNewsId(html: string): string | null {
  const i = html.indexOf("data-news_id=");
  if (i < 0) return null;
  const rest = html.slice(i + "data-news_id=".length).replace(/^["']+/, "");
  const id = /^[0-9]+/.exec(rest);
  return id ? id[0] : null;
}

export function parseMeta(html: string): Meta {
  const $ = cheerio.load(html);
  const one = (css: string): string | null => {
    const el = $(css).first();
    return el.length > 0 ? textOf($, el[0]!) : null;
  };
  const attr = (css: string, name: string): string | null => $(css).first().attr(name) ?? null;

  const facts: Fact[] = [];
  let genres: string[] = [];
  $(".blkdesc p").each((_, p) => {
    const lab = $(p).find(".timpact").first();
    if (lab.length === 0) return;
    const labText = textOf($, lab[0]!);
    const key = labText.replace(/:+$/, "");
    const full = textOf($, p);
    const value = (full.startsWith(labText) ? full.slice(labText.length) : full).trim();
    if (key.toLowerCase() === "жанр") {
      genres = $(p)
        .find("a")
        .map((_, a) => textOf($, a))
        .get();
    }
    if (value) facts.push({ key, value });
  });

  const described = one("[itemprop=description]");
  return {
    title: one("h1") ?? "Без названия",
    original: one("h2.romanji") ?? "",
    poster: attr('meta[property="og:image"]', "content") ?? "",
    description:
      (described !== null ? described.replace(/^(Описание:)+/, "").trim() : null) ??
      attr('meta[property="og:description"]', "content") ??
      "",
    genres,
    facts,
  };
}

/// У гостя страница отдаётся без news_id; на зеркалах без входа он есть сразу.
export function isAuthorized(html: string): boolean {
  return parseNewsId(html) !== null;
}
